### PR TITLE
Added xformers support to Llama

### PR DESCRIPTION
**Pull request by MarkovInequality**
This adds the option to replace llama's default attention with xformer's memory efficient attention.

After some testing, it does not appear that xformers decreases llama's memory footprint all that much nor does it speed up regular inference as much as it does with stable diffusion. However, it does increase the inference times for when you generate with a large context.

Normal Attention
Capture1

Xformers Attention
Capture2

### CODE REVIEW

**Comment by thot-experiment** [comment]
Looks like great work! However, it seems like judging by your screenshot you may have meant to say that it **decreases** inference time? 

> However, it does **increase** the inference times for when you generate with a large context.

Recently IIRC torch2 has implemented this or something similar natively without the need for xformers? I don't know much about the nitty gritty of it so I'm not even sure why I'm bringing it up, but maybe that could be a way further forward?

**Comment by Ph0rk0z** [comment]
Why only llama? Why not others? I will definitely give it a whirl... I need all the help I can get on my old card. Did it work with GPTQ?

File: modules/models.py [hunk 1]
```diff
@@ -179,6 +180,9 @@ def load_model(model_name):
     tokenizer.truncation_side = 'left'
 
     print(f"Loaded the model in {(time.time()-t0):.2f} seconds.")
+
+    llama_attn_hijack.hijack_llama_attention()
```
Comment (Ph0rk0z) [inline modules/models.py:184]: Don't need the if xformers here? We just run it all the time?

**Comment by EyeDeck** [comment]
Some large-context test runs (WSL2, GPTQ-for-LLaMA triton branch, RTX 3090, LLaMA 30B 4-bit act-order groupsize 128):

Control:
>Output generated in 65.74 seconds (3.03 tokens/s, 199 tokens, context 1848)
Output generated in 69.57 seconds (2.86 tokens/s, 199 tokens, context 1848)
Output generated in 63.45 seconds (3.14 tokens/s, 199 tokens, context 1848)
Output generated in 64.79 seconds (3.07 tokens/s, 199 tokens, context 1848)
Output generated in 60.89 seconds (3.27 tokens/s, 199 tokens, context 1848)
Output generated in 62.67 seconds (3.18 tokens/s, 199 tokens, context 1848)
Output generated in 64.70 seconds (3.08 tokens/s, 199 tokens, context 1848)
Output generated in 60.99 seconds (3.26 tokens/s, 199 tokens, context 1848)

Average: 3.105 tokens/s

---

--sdp-attention:
>Output generated in 49.33 seconds (4.03 tokens/s, 199 tokens, context 1848)
Output generated in 55.25 seconds (3.60 tokens/s, 199 tokens, context 1848)
Output generated in 61.90 seconds (3.21 tokens/s, 199 tokens, context 1848)
Output generated in 63.69 seconds (3.12 tokens/s, 199 tokens, context 1848)
Output generated in 63.70 seconds (3.12 tokens/s, 199 tokens, context 1848)
Output generated in 61.64 seconds (3.23 tokens/s, 199 tokens, context 1848)
Output generated in 71.64 seconds (2.78 tokens/s, 199 tokens, context 1848)
Output generated in 61.95 seconds (3.21 tokens/s, 199 tokens, context 1848)

Average: 3.255 tokens/s

---

--xformers:
>Output generated in 55.00 seconds (3.62 tokens/s, 199 tokens, context 1848)
Output generated in 61.08 seconds (3.26 tokens/s, 199 tokens, context 1848)
Output generated in 59.70 seconds (3.33 tokens/s, 199 tokens, context 1848)
Output generated in 57.36 seconds (3.47 tokens/s, 199 tokens, context 1848)
Output generated in 63.45 seconds (3.14 tokens/s, 199 tokens, context 1848)
Output generated in 50.22 seconds (3.96 tokens/s, 199 tokens, context 1848)
Output generated in 56.78 seconds (3.50 tokens/s, 199 tokens, context 1848)
Output generated in 56.92 seconds (3.50 tokens/s, 199 tokens, context 1848)

Average: 3.457 tokens/s

---

Can't measure any VRAM usage difference, all three climb up to 99.8-100% as measured by HWiNFO.
At least in this test, --xformers is about 11% faster than baseline, so it's an improvement.

**Comment by oobabooga** [comment]
That's very impressive. In my tests, there was a decrease in performance with xformers using LLaMA 7b in 16-bit mode:

* Without xformers: Output generated in 6.00 seconds (33.15 tokens/s, 199 tokens, context 16)
* With xformers: Output generated in 6.61 seconds (30.10 tokens/s, 199 tokens, context 16)

But there are many other possibilities to experiment with (8-bit, 4-bit, offloading). 

Thank you for submitting such sophisticated PR.

**Comment by VldmrB** [comment]
I was all excited, seeing as building xformers from source and using it in stable-web-ui increased generation speed by 100%+ for me, but here, after building it from scratch again, it's about the same, probably slightly slower.

I'm on Windows, but that probably doesn't matter.
For those curious, here's the excellent post that describes building xformers from source, on Windows: https://github.com/AUTOMATIC1111/stable-diffusion-webui/issues/2449#issuecomment-1404540735
Maybe there's something that's off and it will be fixed, fingers crossed.

**Comment by EyeDeck** [comment]
I think it's more just that the workload doesn't happen to benefit as much from xformers as some others do. The GPTQ-for-LLaMA guy, qwopqwop200, mentioned that he tried implementing xformers over there once, and he didn't see any benefit either, so he gave up on that code.

**Comment by Ph0rk0z** [comment]
For me it's the same. It didn't really do much one way or another. I had to upgrade pytorch to use it.



```
xformers

7b
Output generated in 3.37 seconds (3.56 tokens/s, 12 tokens, context 207)
Output generated in 2.97 seconds (4.04 tokens/s, 12 tokens, context 207)
Output generated in 2.95 seconds (4.07 tokens/s, 12 tokens, context 207)

7b fp16
Output generated in 2.59 seconds (5.78 tokens/s, 15 tokens, context 418)
Output generated in 2.28 seconds (6.59 tokens/s, 15 tokens, context 418)
Output generated in 2.30 seconds (6.52 tokens/s, 15 tokens, context 418)


30b
Output generated in 8.22 seconds (0.85 tokens/s, 7 tokens, context 207)
Output generated in 7.94 seconds (0.88 tokens/s, 7 tokens, context 207)
Output generated in 7.93 seconds (0.88 tokens/s, 7 tokens, context 207)

Output generated in 31.35 seconds (1.08 tokens/s, 34 tokens, context 418)
Output generated in 16.94 seconds (0.94 tokens/s, 16 tokens, context 418)
Output generated in 19.26 seconds (0.99 tokens/s, 19 tokens, context 418)

30b gptqv2
Output generated in 9.34 seconds (1.93 tokens/s, 18 tokens, context 418)
Output generated in 9.02 seconds (2.00 tokens/s, 18 tokens, context 418)
Output generated in 9.07 seconds (1.98 tokens/s, 18 tokens, context 418



non xformers for GPTQv1 
7b
Output generated in 3.24 seconds (3.71 tokens/s, 12 tokens, context 207)
Output generated in 2.96 seconds (4.05 tokens/s, 12 tokens, context 207)
Output generated in 2.97 seconds (4.04 tokens/s, 12 tokens, context 207)

7b fp16
Output generated in 2.50 seconds (6.01 tokens/s, 15 tokens, context 418)
Output generated in 2.20 seconds (6.83 tokens/s, 15 tokens, context 418)
Output generated in 2.22 seconds (6.77 tokens/s, 15 tokens, context 418)

30b
Output generated in 8.19 seconds (0.85 tokens/s, 7 tokens, context 207)
Output generated in 7.90 seconds (0.89 tokens/s, 7 tokens, context 207)
Output generated in 7.86 seconds (0.89 tokens/s, 7 tokens, context 207)

Output generated in 12.55 seconds (0.80 tokens/s, 10 tokens, context 418)
Output generated in 17.66 seconds (0.96 tokens/s, 17 tokens, context 418)
Output generated in 24.81 seconds (1.05 tokens/s, 26 tokens, context 418)

30b gptqv2
Output generated in 9.13 seconds (1.97 tokens/s, 18 tokens, context 418)
Output generated in 8.84 seconds (2.04 tokens/s, 18 tokens, context 418)
Output generated in 8.88 seconds (2.03 tokens/s, 18 tokens, context 418)

```

**Comment by dronelove** [comment]
can someone explain how to enable xformers? I tried adding --xformers to startwebui.bat but got an error saying "xformers is not defined." thanx

**Comment by thot-experiment** [comment]
> I tried adding --xformers to startwebui.bat but got an error saying "xformers is not defined." 
I have the same error, i've confirmed that xformers is installed inside umamba via `pip list` which returns `xformers               0.0.18`

additionally creating a file `test.py` with one line `import xformers` returns the same error, so this is perhaps some sort of a python issue (not my native language so I'm not sure where to go from here)  `which python` returns the correct path `ooba/installer_files/env/python` while inside umamba cmd

**Comment by Dusterwald** [comment]
I'm currently trying to get xformers to work (mainly in the hope that it won't OOM on 30B 4bit models on my 4090).
Running it on Win 10 with WSL2. I can get the UI to load fine, but on inference it throws the following error:

`File "text-generation-webui/modules/llama_attn_hijack.py", line 48, in xformers_forward
    query_states, key_states = transformers.models.llama.modeling_llama.apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
  File "miniconda3/envs/textgen/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py", line 130, in apply_rotary_pos_emb
    cos = cos[..., offset : q.shape[-2] + offset, :]
TypeError: unsupported operand type(s) for +: 'int' and 'NoneType'`

I've tried updating transformers/xformers/pytorch to the latest, but no luck so far. I'm going to jump into the code and see if I can figure out what is going on, but if anyone has any ideas in the meantime I'd be happy to hear those.

**Comment by thot-experiment** [comment]
Figured I'd try this with a fresh install using the oneclick, and now I'm running into the same issue as (#836) `CUDA extension not installed`

**Comment by Ph0rk0z** [comment]
I never tried SDP... I guess I should :)

**Pull request merged**